pyfr/plugins/turbulence.py

```diff
@@ -0,0 +1,256 @@
+
+# pcg32rxs_m_xs
+def pcg(seed):
+    state = np.uint32(seed)
+    multiplier = np.uint32(747796405)
+    increment = np.uint32(2891336453)
+    mcgmultiplier = np.uint32(277803737)
+    n4 = np.uint32(4)
+    n22 = np.uint32(22)
+    n28 = np.uint32(28)
+    with np.errstate(over='ignore'):
+        while True:
+            ostate = state
+            ostatet = state
+            state = ostatet*multiplier + increment
+            rshift = np.uint32(ostatet >> n28)
+            ostatet ^= ostatet >> (n4 + rshift)
+            ostatet *= mcgmultiplier
+            ostatet ^= ostatet >> n22
+            random = (ostatet >> 8) * 5.9604644775390625e-8
+            yield (ostate, random)
+
+class TurbulencePlugin(BaseSolverPlugin):
+    name = 'turbulence'
+    systems = [ 'ac-navier-stokes', 'navier-stokes']
+    formulations = ['dual', 'std']
+    dimensions = [3]
+
+    def __init__(self, intg, cfgsect):
+        super().__init__(intg, cfgsect)
+
+        self.fdptype = fdptype = intg.backend.fpdtype
+
+        ac = intg.system.name.startswith('ac')
+
+        self.nmvxinit = 6
+        self.nvmxtol = 0.1
+
+        self.tstart = intg.tstart
+        self.tbegin = intg.tcurr
+        self.tnext = intg.tcurr
+        self.tend = intg.tend
+
+        self.seed = self.cfg.getint(cfgsect, 'seed')
+
+        self.eventdtype = [('tinit', fdptype), ('state', np.uint32),
+                           ('ts', fdptype), ('te', fdptype)]
+
+        ti = self.cfg.getfloat(cfgsect, 'turbulence-intensity')
+        sigma = self.cfg.getfloat(cfgsect, 'sigma')
+        self.avgu = avgu = self.cfg.getfloat(cfgsect, 'avg-u')
+        self.ls = ls = self.cfg.getfloat(cfgsect, 'turbulence-length-scale')
+
+        gc = (2*sigma/(math.erf(1/sigma)*math.pi**0.5))**0.5
+        beta1 = -0.5/(sigma*ls)**2
+        beta3 = 0.01*ti*avgu*(gc/sigma)**3
+
+        if not ac:
+            gamma = self.cfg.getfloat('constants', 'gamma')
+            avgrho = self.cfg.getfloat(cfgsect, 'avg-rho')
+            avgmach = self.cfg.getfloat(cfgsect, 'avg-mach')
+            beta2 = avgrho*(gamma - 1)*avgmach**2
+        else:
+            beta2 = 0
+
+        self.ydim = ydim = self.cfg.getfloat(cfgsect, 'y-dim')
+        self.zdim = zdim = self.cfg.getfloat(cfgsect, 'z-dim')
+        
+        self.ymin = ymin = -ydim/2
+        self.zmin = zmin = -zdim/2
+
+        self.nvorts = int(ydim*zdim/(4*ls**2))
+        self.bbox = BoxRegion([-2*ls, ymin-ls, zmin-ls],
+                              [2*ls, ymin+ydim+ls, zmin+zdim+ls])
+        
+        centre = self.cfg.getliteral(cfgsect, 'centre')
+        rax = np.array(self.cfg.getliteral(cfgsect, 'rot-axis'))
+        ran = np.radians(self.cfg.getfloat(cfgsect,'rot-angle'))
+
+        self.shift = shift = np.array(centre)
+        self.rot = rot = (np.cos(ran)*np.eye(3) -
+                          np.sin(ran)*(np.cross(rax, -np.eye(3))) +
+                          (1.0 - np.cos(ran))*rax@rax.T)
```

Comment:
Space around @ no need for `.0`. Also, does the sign from the `np.eye(3)` in the `np.cross` not factor out and cancel with the - sign out front of the `np.sin`?